x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/utils/MlIndexAndAlias.java

```diff
@@ -69,13 +72,38 @@ private MlIndexAndAlias() {}
      * Adds an {@code alias} to that index if it was created,
      * or to the index with the highest suffix if the index did not have to be created.
      * The listener is notified with a {@code boolean} that informs whether the index or the alias were created.
+     * If the index is created, the listener is not called until the index is ready to use via the supplied alias,
+     * so that a method that receives a success response from this method can safely use the index immediately.
      */
     public static void createIndexAndAliasIfNecessary(Client client,
                                                       ClusterState clusterState,
                                                       IndexNameExpressionResolver resolver,
                                                       String indexPatternPrefix,
                                                       String alias,
-                                                      ActionListener<Boolean> listener) {
+                                                      ActionListener<Boolean> finalListener) {
+
+        // If the index and alias were successfully created then wait for the shards of the index that the alias points to be ready
+        ActionListener<Boolean> waitForShardsListener = ActionListener.wrap(
+            created -> {
+                if (created) {
+                    ClusterHealthRequest healthRequest = Requests.clusterHealthRequest(alias)
+                        .waitForNoRelocatingShards(true)
+                        .waitForNoInitializingShards(true);
+                    executeAsyncWithOrigin(
+                        client.threadPool().getThreadContext(),
+                        ML_ORIGIN,
+                        healthRequest,
+                        ActionListener.<ClusterHealthResponse>wrap(
+                            response -> finalListener.onResponse(response.isTimedOut() == false),
+                            finalListener::onFailure),
+                        (request, listener) -> client.admin().cluster().health(request, listener)
```

Comment:
I think you could replace this call with a more concise `client.admin().cluster()::health`.